File: Sources/RMBTBaseTestViewController.m
/*
 * Copyright 2017 appscape gmbh
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#import "RMBTBaseTestViewController.h"
#import "CLLocation+RMBTFormat.h"
#import "RMBTTestRunner.h"
#import "RMBT-Swift.h"

@interface RMBTBaseTestViewController ()<RMBTTestRunnerDelegate> {
    NSUInteger _finishedPercentage;
    NSUInteger _finishedGaugePercentage;
    RMBTTestRunner *_testRunner;
    BOOL _qosPerformed;
    BOOL _qosWillPerformed;
}
@end

// The RMBTBaseTestViewControllerSubclass protocol, which subclasses should implement
// and this define let us emulate this being an "abstract" class. We get warning
// in subclasses about unimplemented methods, and don't need to implement them in the base class:
#define subself ((RMBTBaseTestViewController<RMBTBaseTestViewControllerSubclass>*)self)

@implementation RMBTBaseTestViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _qosWillPerformed = !([RMBTSettings sharedSettings].skipQoS);
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [[UIApplication sharedApplication] setIdleTimerDisabled:YES]; // Disallow turning off the screen
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];

    // Allow turning off the screen again.
    // Note that enabling the idle timer won't reset it, so if the device has alredy been idle the screen will dim
    // immediately. To prevent this, we delay enabling by 5s.
    [[UIApplication sharedApplication] bk_performBlock:^(id sender) {
        [[UIApplication sharedApplication] setIdleTimerDisabled:NO];
    } afterDelay:5.0];
}

- (NSUInteger)percentageAfterPhase:(RMBTTestRunnerPhase)phase {
    switch (phase) {
        case RMBTTestRunnerPhaseNone:
            return 0;
        case RMBTTestRunnerPhaseFetchingTestParams:
        case RMBTTestRunnerPhaseWait:
            return 4;
        case RMBTTestRunnerPhaseInit:
            return 20;
        case RMBTTestRunnerPhaseLatency:
            return 40;
        case RMBTTestRunnerPhaseDown:
        case RMBTTestRunnerPhaseInitUp:  // no visualization for init up
            return 70;
        case RMBTTestRunnerPhaseUp:
            return 100;
        case RMBTTestRunnerPhaseQoS:
            return 100;
        case RMBTTestRunnerPhaseSubmittingTestResult:
            return 100; // also no visualization for submission
      }
}

- (NSUInteger)percentageAfterPhaseWithQos:(RMBTTestRunnerPhase)phase {
    switch (phase) {
        case RMBTTestRunnerPhaseNone:
            return 0;
        case RMBTTestRunnerPhaseFetchingTestParams:
        case RMBTTestRunnerPhaseWait:
            return 4;
        case RMBTTestRunnerPhaseInit:
            return 15;
        case RMBTTestRunnerPhaseLatency:
            return 30;
        case RMBTTestRunnerPhaseDown:
        case RMBTTestRunnerPhaseInitUp:  // no visualization for init up
            return 50;
        case RMBTTestRunnerPhaseUp:
            return 75;
        case RMBTTestRunnerPhaseQoS:
            return 100;
        case RMBTTestRunnerPhaseSubmittingTestResult:
            return 100; // also no visualization for submission
      }
}

- (NSUInteger)gaugePercentageAfterPhase:(RMBTTestRunnerPhase)phase {
    switch (phase) {
        case RMBTTestRunnerPhaseNone:
            return 0;
        case RMBTTestRunnerPhaseFetchingTestParams:
        case RMBTTestRunnerPhaseWait:
            return 4;
        case RMBTTestRunnerPhaseInit:
            return 10;
        case RMBTTestRunnerPhaseLatency:
            return 23;
        case RMBTTestRunnerPhaseDown:
        case RMBTTestRunnerPhaseInitUp:  // no visualization for init up
            return 49;
        case RMBTTestRunnerPhaseUp:
            return 74;
        case RMBTTestRunnerPhaseQoS:
            return 100;
        case RMBTTestRunnerPhaseSubmittingTestResult:
            return 100; // also no visualization for submission
      }
}

- (NSUInteger)percentageForPhase:(RMBTTestRunnerPhase)phase {
    switch (phase) {
        case RMBTTestRunnerPhaseWait:    return 4 /* waiting phase, visualized as init */;
        case RMBTTestRunnerPhaseInit:    return 16 /* waiting phase, visualized as init */;
        case RMBTTestRunnerPhaseLatency: return 20;
        case RMBTTestRunnerPhaseDown:    return 30;
        case RMBTTestRunnerPhaseUp:      return 30;
        case RMBTTestRunnerPhaseQoS:     return 30;
        default: return 0;
    }
}

- (NSUInteger)gaugePercentageForPhase:(RMBTTestRunnerPhase)phase {
    switch (phase) {
        case RMBTTestRunnerPhaseWait:    return 4 /* waiting phase, visualized as init */;
        case RMBTTestRunnerPhaseInit:    return 6 /* waiting phase, visualized as init */;
        case RMBTTestRunnerPhaseLatency: return 13;
        case RMBTTestRunnerPhaseDown:    return 26;
        case RMBTTestRunnerPhaseUp:      return 25;
        default: return 0;
    }
}

- (NSUInteger)gaugePercentageForPhaseWithQos:(RMBTTestRunnerPhase)phase {
    switch (phase) {
        case RMBTTestRunnerPhaseWait:    return 4 /* waiting phase, visualized as init */;
        case RMBTTestRunnerPhaseInit:    return 6 /* waiting phase, visualized as init */;
        case RMBTTestRunnerPhaseLatency: return 13;
        case RMBTTestRunnerPhaseDown:    return 26;
        case RMBTTestRunnerPhaseUp:      return 25;
        case RMBTTestRunnerPhaseQoS:      return 26;
        default: return 0;
    }
}

- (NSUInteger)percentageForPhaseWithQos:(RMBTTestRunnerPhase)phase {
    switch (phase) {
        case RMBTTestRunnerPhaseWait:    return 4;
        case RMBTTestRunnerPhaseInit:    return 11; /* waiting phase, visualized as init */;
        case RMBTTestRunnerPhaseLatency: return 15;
        case RMBTTestRunnerPhaseDown:    return 20;
        case RMBTTestRunnerPhaseUp:      return 25;
        case RMBTTestRunnerPhaseQoS:     return 25;
        default: return 0;
    }
}

- (NSString*)statusStringForPhase:(RMBTTestRunnerPhase)phase {
    switch(phase) {
        case RMBTTestRunnerPhaseNone:
        case RMBTTestRunnerPhaseFetchingTestParams:
            return NSLocalizedString(@"Fetching test parameters", @"Phase status label");
        case RMBTTestRunnerPhaseWait:
            return NSLocalizedString(@"Waiting for test server", @"Phase status label");
        case RMBTTestRunnerPhaseInit:
            return NSLocalizedString(@"Initializing", @"Phase status label");
        case RMBTTestRunnerPhaseLatency:
            return NSLocalizedString(@"Pinging", @"Phase status label");
        case RMBTTestRunnerPhaseDown:
            return NSLocalizedString(@"Download", @"Phase status label");
        case RMBTTestRunnerPhaseInitUp:
            return NSLocalizedString(@"Initializing Upload", @"Phase status label");
        case RMBTTestRunnerPhaseUp:
            return NSLocalizedString(@"Upload", @"Phase status label");
        case RMBTTestRunnerPhaseQoS:
            return NSLocalizedString(@"measurement_qos", @"Phase status label");
        case RMBTTestRunnerPhaseSubmittingTestResult:
            return NSLocalizedString(@"Finalizing", @"Phase status label");
    }
}

- (void)startTestWithExtraParams:(NSDictionary*)extraParams {
    _finishedPercentage = 0;
    _finishedGaugePercentage = 0;
    _qosPerformed = NO;
    [subself onTestUpdatedTotalProgress:0 gaugeProgress:0];
    [subself onTestUpdatedServerName:@"-"];
    [subself onTestUpdatedStatus:@"-"];
    _testRunner = [[RMBTTestRunner alloc] initWithDelegate:self];
    [_testRunner startWithExtraParams:extraParams];
}

- (void)cancelTest {
    [[RMBTControlServer sharedControlServer] cancelAllRequests];
    [_testRunner cancel]; // TODO: move control server to runner
}

- (void)testRunnerDidDetectConnectivity:(RMBTConnectivity*)connectivity {
    [subself onTestUpdatedConnectivity:connectivity];
}

- (void)testRunnerDidDetectLocation:(CLLocation*)location {
    [subself onTestUpdatedLocation:location];
}

- (void)testRunnerDidStartPhase:(RMBTTestRunnerPhase)phase {
    if (phase == RMBTTestRunnerPhaseInit || phase == RMBTTestRunnerPhaseWait) {
        [subself onTestUpdatedServerName:_testRunner.testParams.serverName];
    }
    [subself onTestUpdatedStatus:[self statusStringForPhase:phase]];
    [subself onTestStartedPhase:phase];
}

- (void)testRunnerDidFinishPhase:(RMBTTestRunnerPhase)phase {
    if (_qosWillPerformed) {
        _finishedPercentage = [self percentageAfterPhaseWithQos:phase];
    } else {
        _finishedPercentage = [self percentageAfterPhase:phase];
    }
    _finishedGaugePercentage = [self gaugePercentageAfterPhase:phase];
    
    NSAssert(_finishedPercentage <= 100, @"Invalid percentage");
    [subself onTestUpdatedTotalProgress:_finishedPercentage gaugeProgress:_finishedGaugePercentage];

    if (phase == RMBTTestRunnerPhaseLatency) {
        [subself onTestMeasuredLatency:_testRunner.testResult.medianPingNanos];
    } else if (phase == RMBTTestRunnerPhaseDown) {
        [subself onTestMeasuredDownloadSpeed:_testRunner.testResult.totalDownloadHistory.totalThroughput.kilobitsPerSecond];
    } else if (phase == RMBTTestRunnerPhaseUp) {
        [subself onTestMeasuredUploadSpeed:_testRunner.testResult.totalUploadHistory.totalThroughput.kilobitsPerSecond];
    } else if (phase == RMBTTestRunnerPhaseQoS) {
        _qosPerformed = YES;
    }

    [subself onTestFinishedPhase:phase];
}

- (void)testRunnerDidUpdateProgress:(float)progress inPhase:(RMBTTestRunnerPhase)phase {
    NSUInteger totalPercentage = _finishedPercentage;
    NSUInteger totalGaugePercentage = _finishedGaugePercentage;
    if (_qosWillPerformed) {
        totalPercentage += [self percentageForPhaseWithQos:phase] * progress;
        totalGaugePercentage += [self gaugePercentageForPhaseWithQos:phase] * progress;
    } else {
        totalPercentage += [self percentageForPhase:phase] * progress;
        totalGaugePercentage += [self gaugePercentageForPhase:phase] * progress;
    }
    NSAssert(totalPercentage <= 100, @"Invalid percentage");
    [subself onTestUpdatedTotalProgress:totalPercentage gaugeProgress:totalGaugePercentage];
}

- (void)testRunnerDidMeasureThroughputs:(NSArray*)throughputs inPhase:(RMBTTestRunnerPhase)phase {
    [subself onTestMeasuredTroughputs:throughputs inPhase:phase];
}

- (void)testRunnerDidCompleteWithResult:(RMBTHistoryResult*)result {
    [subself onTestCompletedWithResult:result qos:_qosPerformed];
}

- (void)testRunnerDidCancelTestWithReason:(RMBTTestRunnerCancelReason)cancelReason {
    [subself onTestUpdatedStatus:NSLocalizedString(@"Aborted", @"Test status")];
    [subself onTestCancelledWithReason:cancelReason];
}

- (void)testRunnerQoSDidStartWithGroups:(NSArray*)groups {
    [subself onTestStartedQoSWithGroups:groups];
}

- (void)testRunnerQoSGroup:(RMBTQoSTestGroup*)group didUpdateProgress:(float)progress {
    [subself onTestUpdatedProgress:progress inQoSGroup:group];
}

@end

File: Sources/RMBTBaseTestViewController.swift
//
//  RMBTBaseTestViewController.swift
//  RMBT
//
//  Created by Sergey Glushchenko on 15.12.2021.
//  Copyright © 2021 appscape gmbh. All rights reserved.
//

import UIKit
import CoreLocation

protocol RMBTBaseTestViewControllerSubclass: AnyObject {

    func onTestUpdatedTotalProgress(_ percentage: UInt, gaugeProgress gaugePercentage: UInt)
    
    func onTestUpdatedStatus(_ status: String)
    func onTestUpdatedConnectivity(_ connectivity: RMBTConnectivity)
    func onTestUpdatedLocation(_ location: CLLocation)
    func onTestUpdatedServerName(_ name: String)

    func onTestStartedPhase(_ phase: RMBTTestRunnerPhase)
    func onTestFinishedPhase(_ phase: RMBTTestRunnerPhase)

    func onTestMeasuredLatency(_ nanos: UInt64)
    func onTestMeasuredPings(_ pings: [Ping], in phase: RMBTTestRunnerPhase)
    func onTestMeasuredTroughputs(_ throughputs: [RMBTThroughput], in phase: RMBTTestRunnerPhase)

    func onTestMeasuredDownloadSpeed(_ kbps: UInt32)
    func onTestMeasuredUploadSpeed(_ kbps: UInt32)

    func onTestStartedQoS(with groups: [RMBTQoSTestGroup], and tests: [RMBTQoSTest])
    func onTestUpdatedProgress(_ progress: Float, in group: RMBTQoSTestGroup)
    func onTestUpdatedProgress(_ testProgress: Float, for test: RMBTQoSTest, in group: RMBTQoSTestGroup)
    
    func onTestCompleted(with result: RMBTHistoryResult!, qos qosPerformed: Bool)
    func onTestCancelled(with reason: RMBTTestRunnerCancelReason)
}

// A base class for both regular test view controller and loop view controller, which have different
// display logic. The behaviour is customized by implementing the RMBTBaseTestViewControllerSubclass protocol
class RMBTBaseTestViewController: UIViewController {

    private var finishedPercentage: UInt = 0
    private var finishedGaugePercentage: UInt = 0
    private var testRunner: RMBTTestRunner?
    private var qosPerformed: Bool = false
    private var qosWillPerformed: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()

        qosWillPerformed = RMBTTestRunner.willQoSPerformed()
    }
    
    func testStartTime() -> UInt64 {
        return testRunner?.testResult?.testStartNanos ?? 0
    }
    
    func percentage(after phase: RMBTTestRunnerPhase) -> UInt {
        switch(phase) {
        case .none: return 0
        case .fetchingTestParams, .wait: return 4
        case .`init`: return 20
        case .latency: return 40
        case .down, .initUp:  // no visualization for init up
            return 70
        case .up: return 100
        case .qos: return 100
        case .submittingTestResult:
            return 100 // also no visualization for submission
        }
    }
    
    func percentageAfterPhaseWithQos(_ phase: RMBTTestRunnerPhase) -> UInt {
        switch(phase) {
        case .none: return 0
        case .fetchingTestParams, .wait: return 4
        case .`init`: return 15
        case .latency: return 30
        case .down, .initUp:  // no visualization for init up
            return 50
        case .up: return 75
        case .qos: return 100
        case .submittingTestResult:
            return 100 // also no visualization for submission
        }
    }
    
    func gaugePercentage(after phase: RMBTTestRunnerPhase) -> UInt {
        switch(phase) {
        case .none: return 0
        case .fetchingTestParams, .wait: return 4
        case .`init`: return 10
        case .latency: return 23
        case .down, .initUp:  // no visualization for init up
            return 49
        case .up: return 74
        case .qos: return 100
        case .submittingTestResult:
            return 100 // also no visualization for submission
        }
    }

    func percentage(for phase: RMBTTestRunnerPhase) -> UInt {
        switch (phase) {
        case .wait:    return 4 /* waiting phase, visualized as init */
        case .`init`:    return 16 /* waiting phase, visualized as init */
        case .latency: return 20
        case .down:    return 30
        case .up:      return 30
        case .qos:     return 30
        default: return 0;
        }
    }
    
    func gaugePercentage(for phase: RMBTTestRunnerPhase) -> UInt {
        switch (phase) {
        case .wait:    return 4 /* waiting phase, visualized as init */
        case .`init`:    return 6 /* waiting phase, visualized as init */
        case .latency: return 13
        case .down:    return 26
        case .up:      return 25
        default: return 0
        }
    }

    func percentageForPhaseWithQos(_ phase: RMBTTestRunnerPhase) -> UInt {
        switch (phase) {
        case .wait:    return 4 /* waiting phase, visualized as init */
        case .`init`:    return 11 /* waiting phase, visualized as init */
        case .latency: return 15
        case .down:    return 20
        case .up:      return 25
        case .qos:     return 25
        default: return 0;
        }
    }
    
    func gaugePercentageForPhaseWithQos(_ phase: RMBTTestRunnerPhase) -> UInt {
        switch (phase) {
        case .wait:    return 4 /* waiting phase, visualized as init */
        case .`init`:    return 6 /* waiting phase, visualized as init */
        case .latency: return 13
        case .down:    return 26
        case .up:      return 25
        case .qos:      return 26
        default: return 0;
        }
    }
    
    func statusString(for phase: RMBTTestRunnerPhase) -> String {
        switch(phase) {
        case .none, .fetchingTestParams:
            return NSLocalizedString("Fetching test parameters", comment: "Phase status label")
        case .wait:
            return NSLocalizedString("Waiting for test server", comment: "Phase status label")
        case .`init`:
            return NSLocalizedString("Initializing", comment: "Phase status label")
        case .latency:
            return NSLocalizedString("Pinging", comment: "Phase status label")
        case .down:
            return NSLocalizedString("Download", comment: "Phase status label");
        case .initUp:
                return NSLocalizedString("Initializing Upload", comment: "Phase status label")
        case .up:
            return NSLocalizedString("Upload", comment: "Phase status label")
        case .qos:
            return NSLocalizedString("measurement_qos", comment: "Phase status label");
        case .submittingTestResult:
            return NSLocalizedString("Finalizing", comment: "Phase status label")
        }
    }
    
    func startTest(with extraParams: [String: Any]?) {
        UIApplication.shared.isIdleTimerDisabled = true // Disallow turning off the screen
        finishedPercentage = 0
        finishedGaugePercentage = 0
        qosPerformed = false
        if let subself = self as? RMBTBaseTestViewControllerSubclass {
            subself.onTestUpdatedTotalProgress(0, gaugeProgress: 0)
            subself.onTestUpdatedServerName("-")
            subself.onTestUpdatedStatus("-")
        }
        testRunner = RMBTTestRunner(delegate: self)
        testRunner?.start(with: extraParams)
    }
    
    func cancelTest() {
        RMBTControlServer.shared.cancelAllRequests()
        testRunner?.cancel() // TODO: move control server to runner
        finishTest()
    }
    
    func finishTest() {
        UIApplication.shared.isIdleTimerDisabled = false
        RMBTSettings.shared.activeMeasurementId = nil
    }
}

extension RMBTBaseTestViewController: RMBTTestRunnerDelegate {
    func testRunnerDidStart(_ phase: RMBTTestRunnerPhase) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        
        if (phase == .`init` || phase == .wait) {
            subself.onTestUpdatedServerName(testRunner?.testParams?.serverName ?? "")
        }
        subself.onTestUpdatedStatus(self.statusString(for: phase))
        subself.onTestStartedPhase(phase)
    }
    
    func testRunnerDidFinish(_ phase: RMBTTestRunnerPhase) {
        guard let testRunner = self.testRunner else { return }
        if (qosWillPerformed) {
            finishedPercentage = self.percentageAfterPhaseWithQos(phase)
        } else {
            finishedPercentage = self.percentage(after: phase)
        }
        finishedGaugePercentage = self.gaugePercentage(after: phase)
        
        assert(finishedPercentage <= 100, "Invalid percentage")
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        
        subself.onTestUpdatedTotalProgress(finishedPercentage, gaugeProgress: finishedGaugePercentage)

        if (phase == .latency) {
            subself.onTestMeasuredLatency(testRunner.testResult?.medianPingNanos ?? 0)
        } else if (phase == .down) {
            subself.onTestMeasuredDownloadSpeed(UInt32(testRunner.testResult?.totalDownloadHistory.totalThroughput.kilobitsPerSecond() ?? 0))
        } else if (phase == .up) {
            subself.onTestMeasuredUploadSpeed(UInt32(testRunner.testResult?.totalUploadHistory.totalThroughput.kilobitsPerSecond() ?? 0))
        } else if (phase == .qos) {
            qosPerformed = true
        }

        subself.onTestFinishedPhase(phase)
    }
    
    func testRunnerDidUpdateProgress(_ progress: Float, in phase: RMBTTestRunnerPhase) {
        var totalPercentage = Float(finishedPercentage)
        var totalGaugePercentage = Float(finishedGaugePercentage)
        if (qosWillPerformed) {
            totalPercentage += Float(self.percentageForPhaseWithQos(phase)) * progress
            totalGaugePercentage += Float(self.gaugePercentageForPhaseWithQos(phase)) * progress
        } else {
            totalPercentage += Float(self.percentage(for: phase)) * progress
            totalGaugePercentage += Float(self.gaugePercentage(for: phase)) * progress
        }
        
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestUpdatedTotalProgress(UInt(totalPercentage), gaugeProgress: UInt( totalGaugePercentage))
    }
    
    func testRunnerQoSGroup(_ group: RMBTQoSTestGroup, didUpdateProgress progress: Float, in test: RMBTQoSTest) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestUpdatedProgress(progress, for: test, in: group)
    }
    
    func testRunnerDidMeasurePings(_ pings: [Ping], in phase: RMBTTestRunnerPhase) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestMeasuredPings(pings, in: phase)
    }
    
    func testRunnerDidMeasureThroughputs(_ throughputs: [RMBTThroughput], in phase: RMBTTestRunnerPhase) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestMeasuredTroughputs(throughputs, in: phase)
    }
    
    func testRunnerDidDetectConnectivity(_ connectivity: RMBTConnectivity) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestUpdatedConnectivity(connectivity)
    }
    
    func testRunnerDidDetectLocation(_ location: CLLocation) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestUpdatedLocation(location)
    }
    
    func testRunnerDidCompleteWithResult(_ result: RMBTHistoryResult) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestCompleted(with: result, qos: qosPerformed)
    }
    
    func testRunnerDidCancelTestWithReason(_ cancelReason: RMBTTestRunnerCancelReason) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestUpdatedStatus(NSLocalizedString("Aborted", comment: "Test status"))
        subself.onTestCancelled(with: cancelReason)
    }
    
    func testRunnerQoSDidStartWithGroups(_ groups: [RMBTQoSTestGroup], _ tests: [RMBTQoSTest]) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestStartedQoS(with: groups, and: tests)
    }
    
    func testRunnerQoSGroup(_ group: RMBTQoSTestGroup, didUpdateProgress progress: Float) {
        guard let subself = self as? RMBTBaseTestViewControllerSubclass else { return }
        subself.onTestUpdatedProgress(progress, in: group)
    }
    
    
}
